File: riveredge-frontend/src/apps/kuaizhizao/services/sales-contract.ts
/**

 * 销售合同 API

 */



import { apiRequest } from '../../../services/api';

import type { DocumentPrintApiResult } from '../../../utils/printResponseHelpers';

import type { BackendLifecycle } from '../utils/backendLifecycle';
import type { SalesContractTermSnapshot } from './sales-contract-term';



export interface SalesContractItem {

  id?: number;

  material_id: number;

  material_code: string;

  material_name: string;

  material_spec?: string;

  material_unit: string;

  contract_quantity: number;

  released_quantity?: number;

  unit_price: number;

  tax_rate?: number;

  total_amount: number;

  delivery_date?: string;

  notes?: string;

}



export interface SalesContractMilestone {

  id?: number;

  milestone_name: string;

  planned_date: string;

  planned_amount?: number;

  planned_ratio?: number;

  billing_trigger?: string;

  status?: string;

  receivable_id?: number;

  receivable_code?: string;

  notes?: string;

}



export interface SalesContract {

  id?: number;

  contract_code?: string;

  contract_type?: 'single' | 'framework' | string;

  /** 是否录入明细：false=金额总框（无产品行，手填总金额） */
  enter_line_items?: boolean;

  customer_id?: number;

  customer_name?: string;

  customer_contact?: string;

  customer_phone?: string;

  contract_date?: string;

  valid_from?: string;

  valid_to?: string;

  total_quantity?: number;

  total_amount?: number;

  /** 整单优惠金额（从价税合计扣减，不改明细单价） */
  discount_amount?: number;

  released_quantity?: number;

  released_sales_order_codes?: string[];

  released_amount?: number;

  remaining_quantity?: number;

  remaining_amount?: number;

  price_type?: string;

  currency_code?: string;

  status?: string;

  review_status?: string;

  quotation_id?: number;

  quotation_code?: string;

  payment_terms?: string;

  shipping_address?: string;

  shipping_method?: string;

  salesman_id?: number;

  salesman_name?: string;

  notes?: string;

  term_group_id?: number;

  term_group_name?: string;

  contract_terms?: SalesContractTermSnapshot[];

  items?: SalesContractItem[];

  milestones?: SalesContractMilestone[];

  lifecycle?: BackendLifecycle;

  capabilities?: SalesContractCapabilities;

  created_by?: number;

  created_by_name?: string;

  updated_by?: number;

  updated_by_name?: string;

  created_at?: string;

  updated_at?: string;

}

export interface ActionCapability {
  allowed: boolean;
  reason?: string | null;
}

export interface SalesContractCapabilities {
  update?: ActionCapability;
  delete?: ActionCapability;
  submit?: ActionCapability;
  withdraw_submit?: ActionCapability;
  approve?: ActionCapability;
  reject?: ActionCapability;
  revoke_approval?: ActionCapability;
  push_to_sales_order?: ActionCapability;
  push_to_work_order?: ActionCapability;
  print?: ActionCapability;
  close?: ActionCapability;
  create_change?: ActionCapability;
}



export interface SalesContractAlert {

  alert_type: string;

  contract_id: number;

  contract_code: string;

  customer_name: string;

  message: string;

  severity: string;

  due_date?: string;

  milestone_id?: number;

}



/** 合同预警 feed 行唯一 key（同一合同可有多条里程碑逾期） */
export function salesContractAlertFeedId(alert: SalesContractAlert): string {
  if (alert.milestone_id != null) {
    return `${alert.alert_type}-${alert.contract_id}-${alert.milestone_id}`;
  }
  if (alert.alert_type === 'milestone_overdue') {
    return `${alert.alert_type}-${alert.contract_id}-${alert.due_date ?? alert.message}`;
  }
  return `${alert.alert_type}-${alert.contract_id}`;
}



export interface SalesContractExecutionSummary {

  contract_id: number;

  contract_code: string;

  contract_type: string;

  customer_name: string;

  total_amount: number;

  released_amount: number;

  remaining_amount: number;

  valid_to?: string;

  status: string;

}



export interface SalesContractChange {

  id: number;

  change_code: string;

  contract_id: number;

  contract_code: string;

  change_type: string;

  status: string;

  review_status?: string;

  delta_amount: number;

  new_valid_to?: string;

  new_total_amount?: number;

  reason?: string;

  created_at?: string;

  updated_at?: string;

}



export interface SalesContractPaymentSummary {

  contract_id: number;

  contract_code: string;

  total_amount: number;

  planned_milestone_amount: number;

  invoiced_amount: number;

  collected_amount: number;

  pending_amount: number;

  milestones: Array<{

    id: number;

    milestone_name: string;

    planned_date?: string;

    planned_amount?: number;

    status?: string;

    receivable_id?: number;

    receivable_code?: string;

  }>;

}



export interface ConvertToOrderPayload {

  selected_item_ids?: number[];

  release_lines?: { item_id: number; release_quantity: number }[];

}

export interface PushToWorkOrderFromContractPayload {
  selected_item_ids?: number[];
  release_lines?: { item_id: number; release_quantity: number }[];
  push_mode?: 'draft' | 'confirm';
  work_order_granularity?: 'grouped' | 'peer_group';
}

export interface SalesContractPushPreviewItem {
  item_id: number;
  material_id?: number;
  material_code?: string;
  material_name?: string;
  material_unit?: string;
  quantity: number;
  pushed_quantity: number;
  max_push_quantity: number;
  delivery_date?: string | null;
  blocking_issues?: string[];
  suggested_action?: string;
  source_type?: string;
}

export interface SalesContractPushPreviewResponse {
  target_type: 'sales_order' | 'work_order';
  summary: string;
  items: SalesContractPushPreviewItem[];
  has_blocking_issues?: boolean;
  blocking_reason?: string | null;
  push_mode_default?: 'draft' | 'confirm' | string;
  tip?: string;
}

export interface PullSalesContractFromQuotationResponse {
  success: boolean;
  message: string;
  source_type: 'quotation';
  source_id: number;
  sales_contract: SalesContract;
  quotation: {
    id?: number;
    quotation_code?: string;
    status?: string;
    [key: string]: any;
  };
}



export interface SalesContractListParams extends Record<string, unknown> {
  skip?: number;
  limit?: number;
  keyword?: string;
  contract_code?: string;
  status?: string;
  contract_type?: string;
  customer_id?: number;
  start_date?: string;
  end_date?: string;
  order_by?: string;
  /** 明细表格视图：附带合同明细 */
  include_items?: boolean;
}



const BASE = '/apps/kuaizhizao/sales-contracts';



export const salesContractApi = {

  list: (params?: SalesContractListParams) =>

    apiRequest<{ items: SalesContract[]; total: number }>(BASE, { params }),



  get: (id: number, includeItems = true) =>

    apiRequest<SalesContract>(`${BASE}/${id}`, { params: { include_items: includeItems } }),



  create: (data: Partial<SalesContract>, autoSubmit = false) =>

    apiRequest<SalesContract>(BASE, { method: 'POST', params: { auto_submit: autoSubmit }, data }),



  update: (id: number, data: Partial<SalesContract>) =>

    apiRequest<SalesContract>(`${BASE}/${id}`, { method: 'PUT', data }),



  remove: (id: number) => apiRequest(`${BASE}/${id}`, { method: 'DELETE' }),



  submit: (id: number) => apiRequest<SalesContract>(`${BASE}/${id}/submit`, { method: 'POST' }),



  approve: (id: number, review_remarks?: string) =>

    apiRequest<SalesContract>(`${BASE}/${id}/approve`, { method: 'POST', data: { review_remarks } }),



  reject: (id: number, review_remarks?: string) =>

    apiRequest<SalesContract>(`${BASE}/${id}/reject`, { method: 'POST', data: { review_remarks } }),



  close: (id: number, reason?: string) =>

    apiRequest<SalesContract>(`${BASE}/${id}/close`, { method: 'POST', data: { reason } }),



  convertFromQuotation: (quotationId: number, contractType = 'single') =>

    apiRequest<SalesContract>(`${BASE}/from-quotation/${quotationId}`, {

      method: 'POST',

      params: { contract_type: contractType },

    }),

  convertFromSalesOrder: (salesOrderId: number) =>
    apiRequest<SalesContract>(`${BASE}/from-sales-order/${salesOrderId}`, {
      method: 'POST',
    }),

  pullSalesContractFromQuotation: async (quotationId: number, contractType = 'single') => {
    const salesContract = await apiRequest<SalesContract>(`${BASE}/from-quotation/${quotationId}`, {
      method: 'POST',
      params: { contract_type: contractType },
    });
    return {
      success: true,
      message: '已从报价单创建销售合同',
      source_type: 'quotation' as const,
      source_id: quotationId,
      sales_contract: salesContract,
      quotation: { id: quotationId },
    } satisfies PullSalesContractFromQuotationResponse;
  },



  convertToOrder: (id: number, payload?: ConvertToOrderPayload | null) =>

    apiRequest<{ sales_order: Record<string, unknown>; contract: SalesContract }>(

      `${BASE}/${id}/convert-to-order`,

      { method: 'POST', data: payload ?? {} },

    ),

  previewPushToSalesOrder: (id: number) =>
    apiRequest<SalesContractPushPreviewResponse>(`${BASE}/${id}/push-to-sales-order/preview`, {
      method: 'GET',
    }),

  previewPushToWorkOrder: (id: number) =>
    apiRequest<SalesContractPushPreviewResponse>(`${BASE}/${id}/push-to-work-order/preview`, {
      method: 'GET',
    }),

  pushToWorkOrder: (id: number, payload?: PushToWorkOrderFromContractPayload | null) =>
    apiRequest<{
      success: boolean;
      message: string;
      target_documents?: { type: string; id: number; code: string }[];
    }>(`${BASE}/${id}/push-to-work-order`, { method: 'POST', data: payload ?? {} }),



  paymentSummary: (id: number) => apiRequest<SalesContractPaymentSummary>(`${BASE}/${id}/payment-summary`),



  listChanges: (contractId: number, params?: { skip?: number; limit?: number }) =>

    apiRequest<SalesContractChange[]>(`${BASE}/${contractId}/changes`, { params }),



  createChange: (contractId: number, data: Partial<SalesContractChange>) =>

    apiRequest<SalesContractChange>(`${BASE}/${contractId}/changes`, { method: 'POST', data }),



  submitChange: (changeId: number) =>

    apiRequest<SalesContractChange>(`${BASE}/changes/${changeId}/submit`, { method: 'POST' }),



  approveChange: (changeId: number) =>

    apiRequest<SalesContractChange>(`${BASE}/changes/${changeId}/approve`, { method: 'POST' }),



  rejectChange: (changeId: number) =>

    apiRequest<SalesContractChange>(`${BASE}/changes/${changeId}/reject`, { method: 'POST' }),



  generateMilestoneReceivable: (contractId: number, milestoneId: number) =>

    apiRequest(`${BASE}/${contractId}/milestones/${milestoneId}/generate-receivable`, { method: 'POST' }),



  listAlerts: () => apiRequest<SalesContractAlert[]>(`${BASE}/alerts`),



  executionSummary: () => apiRequest<SalesContractExecutionSummary[]>(`${BASE}/execution-summary`),

  withdraw: (id: number) =>
    apiRequest<SalesContract>(`${BASE}/${id}/withdraw`, { method: 'POST' }),

  revokeReview: (id: number) =>
    apiRequest<SalesContract>(`${BASE}/${id}/revoke-review`, { method: 'POST' }),

  print: (
    id: number,
    options?: {
      templateUuid?: string;
      outputFormat?: 'html' | 'pdf';
      responseFormat?: 'json' | 'html';
    },
  ) =>
    apiRequest<DocumentPrintApiResult>(
      `${BASE}/${id}/print`,
      {
        method: 'GET',
        params: {
          template_uuid: options?.templateUuid,
          output_format: options?.outputFormat ?? 'pdf',
          response_format: options?.responseFormat ?? 'json',
        },
      },
    ),

  getPrintVariables: (id: number) =>
    apiRequest<{ success: boolean; variables: Record<string, unknown> }>(
      `${BASE}/${id}/print-variables`,
      { method: 'GET' },
    ),
};



export default salesContractApi;

File: riveredge-backend/src/apps/kuaizhizao/schemas/sales_contract.py
"""
销售合同 Pydantic Schemas
"""

from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional
from pydantic import BaseModel, Field
from apps.kuaizhizao.services.document_action_policy.types import SalesContractCapabilities
from core.schemas.base import BaseSchema

from apps.kuaizhizao.schemas.sales_contract_term import SalesContractTermSnapshot
from apps.kuaizhizao.constants.price_type import DEFAULT_SALES_PRICE_TYPE


class SalesContractItemCreate(BaseModel):
    material_id: int
    material_code: str
    material_name: str
    material_spec: Optional[str] = None
    material_unit: str
    contract_quantity: Decimal
    unit_price: Decimal
    tax_rate: Optional[Decimal] = Decimal("0")
    total_amount: Decimal
    variant_attributes: Optional[dict] = None
    delivery_date: Optional[date] = None
    notes: Optional[str] = None


class SalesContractItemResponse(SalesContractItemCreate):
    id: int
    uuid: str
    tenant_id: int
    contract_id: int
    released_quantity: Decimal = Decimal("0")
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SalesContractMilestoneCreate(BaseModel):
    milestone_name: str = Field(..., max_length=200)
    planned_date: date
    planned_amount: Decimal = Decimal("0")
    planned_ratio: Optional[Decimal] = None
    billing_trigger: str = "milestone"
    notes: Optional[str] = None


class SalesContractMilestoneResponse(SalesContractMilestoneCreate):
    id: int
    uuid: str
    tenant_id: int
    contract_id: int
    status: str
    receivable_id: Optional[int] = None
    receivable_code: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SalesContractCreate(BaseModel):
    contract_code: Optional[str] = Field(None, max_length=50, description="合同编码（自动生成，可手工填写）")
    contract_type: str = Field(default="single", description="single / framework")
    customer_id: int
    customer_name: str
    customer_contact: Optional[str] = None
    customer_phone: Optional[str] = None
    contract_date: date
    valid_from: Optional[date] = None
    valid_to: Optional[date] = None
    price_type: str = DEFAULT_SALES_PRICE_TYPE
    currency_code: Optional[str] = "CNY"
    salesman_id: Optional[int] = None
    salesman_name: Optional[str] = None
    shipping_address: Optional[str] = None
    shipping_method: Optional[str] = None
    payment_terms: Optional[str] = None
    term_group_id: Optional[int] = None
    contract_terms: Optional[List[SalesContractTermSnapshot]] = None
    quotation_id: Optional[int] = None
    discount_amount: Decimal = Decimal("0")
    notes: Optional[str] = None
    attachments: Optional[list] = None
    items: List[SalesContractItemCreate] = Field(default_factory=list)
    milestones: List[SalesContractMilestoneCreate] = Field(default_factory=list)


class SalesContractUpdate(BaseModel):
    contract_type: Optional[str] = None
    customer_id: Optional[int] = None
    customer_name: Optional[str] = None
    customer_contact: Optional[str] = None
    customer_phone: Optional[str] = None
    contract_date: Optional[date] = None
    valid_from: Optional[date] = None
    valid_to: Optional[date] = None
    price_type: Optional[str] = None
    currency_code: Optional[str] = None
    salesman_id: Optional[int] = None
    salesman_name: Optional[str] = None
    shipping_address: Optional[str] = None
    shipping_method: Optional[str] = None
    payment_terms: Optional[str] = None
    term_group_id: Optional[int] = None
    contract_terms: Optional[List[SalesContractTermSnapshot]] = None
    discount_amount: Optional[Decimal] = None
    notes: Optional[str] = None
    attachments: Optional[list] = None
    items: Optional[List[SalesContractItemCreate]] = None
    milestones: Optional[List[SalesContractMilestoneCreate]] = None


class SalesContractResponse(BaseSchema):
    id: int
    uuid: str
    tenant_id: int
    contract_code: str
    contract_type: str
    party_type: str
    customer_id: int
    customer_name: str
    customer_contact: Optional[str] = None
    customer_phone: Optional[str] = None
    contract_date: date
    valid_from: Optional[date] = None
    valid_to: Optional[date] = None
    total_quantity: Decimal
    total_amount: Decimal
    discount_amount: Decimal = Decimal("0")
    released_quantity: Decimal
    released_amount: Decimal
    released_sales_order_codes: List[str] = Field(
        default_factory=list,
        description="已从合同释放的销售订单编码（列表 hover 用）",
    )
    remaining_quantity: Optional[Decimal] = None
    remaining_amount: Optional[Decimal] = None
    price_type: str
    currency_code: Optional[str] = None
    status: str
    review_status: str
    reviewer_id: Optional[int] = None
    reviewer_name: Optional[str] = None
    review_time: Optional[datetime] = None
    review_remarks: Optional[str] = None
    salesman_id: Optional[int] = None
    salesman_name: Optional[str] = None
    shipping_address: Optional[str] = None
    shipping_method: Optional[str] = None
    payment_terms: Optional[str] = None
    term_group_id: Optional[int] = None
    term_group_name: Optional[str] = None
    contract_terms: Optional[List[SalesContractTermSnapshot]] = None
    quotation_id: Optional[int] = None
    quotation_code: Optional[str] = None
    root_contract_id: Optional[int] = None
    version_no: int = 1
    previous_contract_id: Optional[int] = None
    notes: Optional[str] = None
    attachments: Optional[list] = None
    is_active: bool = True
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    items: Optional[List[SalesContractItemResponse]] = None
    milestones: Optional[List[SalesContractMilestoneResponse]] = None
    lifecycle: Optional[dict] = None
    capabilities: Optional[SalesContractCapabilities] = Field(
        None,
        description="业务态动作 capabilities（不含 RBAC）",
    )


class SalesContractListResponse(BaseModel):
    items: List[SalesContractResponse]
    total: int


class SalesContractReviewAction(BaseModel):
    review_remarks: Optional[str] = None


class SalesContractReleaseLine(BaseModel):
    item_id: int = Field(..., description="合同明细ID")
    release_quantity: Decimal = Field(..., gt=0, description="本次释放数量")


class SalesContractConvertToOrderRequest(BaseModel):
    selected_item_ids: Optional[List[int]] = Field(None, description="兼容：选中明细全量释放剩余")
    release_lines: Optional[List[SalesContractReleaseLine]] = Field(
        None, description="按行指定释放数量"
    )


class SalesContractPushToWorkOrderRequest(BaseModel):
    selected_item_ids: Optional[List[int]] = Field(
        None, description="选中合同明细（未传则按 release_lines 或全部可释放）"
    )
    release_lines: Optional[List[SalesContractReleaseLine]] = Field(
        None, description="按行指定本次下推数量"
    )
    push_mode: Optional[str] = Field(
        None, description="下推模式：draft/confirm（不传则使用租户默认）"
    )
    work_order_granularity: Optional[str] = Field(
        None, description="工单粒度：grouped/per_unit（不传默认 grouped）"
    )


class SalesContractChangeCreate(BaseModel):
    change_type: str = "amendment"
    delta_amount: Decimal = Decimal("0")
    new_valid_to: Optional[date] = None
    new_total_amount: Optional[Decimal] = None
    reason: Optional[str] = None


class SalesContractChangeResponse(BaseModel):
    id: int
    uuid: str
    tenant_id: int
    change_code: str
    contract_id: int
    contract_code: str
    change_type: str
    status: str
    review_status: str
    delta_amount: Decimal
    new_valid_to: Optional[date] = None
    new_total_amount: Optional[Decimal] = None
    reason: Optional[str] = None
    new_contract_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SalesContractAlertItem(BaseModel):
    alert_type: str
    contract_id: int
    contract_code: str
    customer_name: str
    message: str
    severity: str = "medium"
    due_date: Optional[date] = None


class SalesContractExecutionSummary(BaseModel):
    contract_id: int
    contract_code: str
    contract_type: str
    customer_name: str
    total_amount: Decimal
    released_amount: Decimal
    remaining_amount: Decimal
    valid_to: Optional[date] = None
    status: str
